src/v2.rs

```diff
@@ -1060,52 +1315,190 @@ impl EszipV2 {
+                let in_memory_fs = InMemoryFs::default();
+                let mut files = vec![(
+                  // files must be namespaced due to thread-local cache in deno_node::package_json.rs
+                  format!("/{}/package.json", nv_ref.nv().name),
+                  String::from_utf8_lossy(&package_json.source).into_owned(),
+                )];
+                files.extend(package.modules.keys().filter_map(|nv_ref| {
+                  Some((
+                    format!("/{}/{}", nv_ref.nv().name, nv_ref.sub_path()?),
+                    // The source code is irrelevant for the entrypoint resolution
+                    String::new(),
+                  ))
+                }));
+                in_memory_fs.setup_text_files(files);
+
+                #[derive(Debug)]
+                struct NoopNpmResolver;
+                impl NpmResolver for NoopNpmResolver {
+                  fn resolve_package_folder_from_package(
+                    &self,
+                    _: &str,
+                    _: &ModuleSpecifier,
+                  ) -> Result<std::path::PathBuf, anyhow::Error>
+                  {
+                    unreachable!()
+                  }
+
+                  fn in_npm_package(&self, _: &ModuleSpecifier) -> bool {
+                    true
+                  }
+
+                  fn ensure_read_permission(
+                    &self,
+                    _: &mut dyn deno_node::NodePermissions,
+                    _: &std::path::Path,
+                  ) -> Result<(), anyhow::Error> {
+                    Ok(())
+                  }
+                }
+                let node_resolver = deno_node::NodeResolver::new(
+                  Rc::new(in_memory_fs),
+                  Rc::new(NoopNpmResolver),
+                );
+                let resolution = node_resolver
+                  .resolve_package_subpath_from_deno_module(
+                    Path::new(&format!("/{}", nv_ref.nv().name)),
+                    None,
+                    &npm_module.specifier,
+                    deno_node::NodeResolutionMode::Execution,
+                  )
+                  .ok()
+                  .flatten();
+                if let Some(resolution) = resolution {
+                  let sub_path = resolution
+                    .into_url()
+                    .path()
+                    .strip_prefix(&format!("/{}/", nv_ref.nv().name))
+                    .unwrap()
+                    .to_string();
+
+                  // Update nv_ref with the resolved sub_path (npm:package => npm:package/entrypoint)
+                  nv_ref = NpmPackageNvReference::new(PackageNvReference {
+                    nv: nv_ref.into_inner().nv,
+                    sub_path: Some(sub_path),
+                  });
+                }
+              }
+            }
+            let module = package.take_module(&nv_ref);
+            if let Some(module) = module {
+              modules.insert(
```

Comment:
It looks like this only inserts the entrypoint, but the remaining modules in the package will be inserted at the end? Also, I think we need to figure out what packages this package depends on and ensure those are included otherwise they'll just be at the end of the eszip as well?

Overall, I think it would be good to talk to @lucacasonato about this PR and ask him review it because he understand the deploy side of things.

(I still feel like we can just get an npm snapshot and then order the packages' cjs/mjs/js files in the eszip based on that, which would also benefit cjs packages. Alternatively, some way to modify an eszip after the first load where the first load reports the order the modules should be ordered in and then the eszip gets modified to reflect that)

Reply:
> It looks like this only inserts the entrypoint, but the remaining modules in the package will be inserted at the end? Also, I think we need to figure out what packages this package depends on and ensure those are included otherwise they'll just be at the end of the eszip as well?

It inserts each npm module in the `ModuleGraph`. If the module's `PackageNvReference` in the `ModuleGraph` is `npm:package@xx.xx.xx` (ie no sub_path), we resolve the entrypoint and load that. Only npm modules not in the `ModuleGraph` are at the end of the eszip. See [this test](https://github.com/denoland/eszip/pull/192/files/54bb954d1b3eaf242b95b87928094265a36117a9#diff-ff42aba765ab96f68cb209933964efa01404f135126af4e1cfe45d5e7c1fd0e9R3184-R3201) for a comprehensive visualization of the result.

> Overall, I think it would be good to talk to @lucacasonato about this PR and ask him review it because he understand the deploy side of things.

Sure! I understand this might be a bit controvertial, and I am happy to redesign it as necessary to resolve any concern. This is just the implementation I came up with that delivers an eszip ordered almost exactly as the order *all* modules are loaded by Deno when booting up, so any alternative implementation should be able to deliver the same eszip.

> I still feel like we can just get an npm snapshot and then order the packages' cjs/mjs/js files in the eszip based on that, which would also benefit cjs packages. 

That could be an evolution of what is here implemented. What I implemented is a mechanism to provide the source code of the npm packages into the eszip, as the `ModuleGraph` does not include it (and the `ModuleGraph` does not include all the npm packages that we want to include in the eszip, namely the cjs). If I'm not mistaken, what you propose is to use the npm snapshot to order the tail of cjs modules. That would be good indeed! I should do that before setting this PR ready to review

> Alternatively, some way to modify an eszip after the first load where the first load reports the order the modules should be ordered in and then the eszip gets modified to reflect that)

What would be the benefit comapred to what I propose here? Sounds a bit more convoluted.

Reply:
@lucacasonato if you review this, keep in mind that @dsherret and I have already discussed OOB about not pulling deno_node and deno_fs dependencies into eszip. The most probably plan is to move the package subpath resolution to the client code (Deploy and others) and provide it as a property of `FromGraphNpmPackage`.